ps.c

```
@@ -0,0 +1,110 @@
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <errno.h>
+
+#define SPACES_TO_CMDLINE 20
+#define MAX_PID_NAME_LENGTH 6 // 32768 (plus slash)
+#define PROC_LENGTH 6 // /proc + space for null terminator
+#define CMDLINE_LENGTH 8
+#define STAT_LENGTH 5
+#define MAX_CMDLINE_LENGTH 70 // Just the max I want to print so it looks good :D
+#define MAX_PROC_NAME 17 // comm[16] in kernel + null terminator
+
+struct ProcessInfo {
+	int pid;
+	char name[MAX_PROC_NAME]; 
+	char cmd_line[MAX_CMDLINE_LENGTH];
+} ;
+
+int main(int argc, char *argv[]){
+	struct dirent *de;
+	DIR *dr = opendir("/proc");
+	int pid;
+	
+	printf("PID\tNAME%-20sCMDLINE\n", " ");
+
+	if (dr == NULL){
+		perror("Couldn't open dir");
+		return 1;
+	}
+
+	while ((de = readdir(dr)) != NULL){
+		pid = atoi(de->d_name);
+		if(pid != 0){
+			print_line(pid);
+		}
+	}
+}
+
+int print_line(int pid){
+	struct ProcessInfo p;
+	p.pid = pid;
+	get_name(&p);
+	get_cmdline(&p);
```

You ignore errors returned by these calls